### PR TITLE
ISSUE-20: implements sig_geterrorstr

### LINKED ISSUE #20: ISSUE-20: implement opendkim.sig_geterrorstr()
Implement `opendkim.sig_geterrorstr()` to get the error string for a signature failure.

- [ ] code
- [ ] tests
- [ ] documentation

For a good example of the above items see: https://github.com/godsflaw/node-opendkim/pull/33
Note that the first commit title should start with `ISSUE-20:`

more information:
http://www.opendkim.org/libopendkim/dkim_sig_geterrorstr.html

[Bountysource](https://www.bountysource.com/issues/50011646-issue-20-implement-opendkim-sig_geterrorstr?utm_source=50011646&utm_medium=shield&utm_campaign=ISSUE_BADGE)
https://ethlance.com/#/job/224

### PR DESCRIPTION
**Pull request by desttinghim**
# Description
This PR implements `sig_geterrorstr()`. This function must be passed an error code from `sig_geterror()` , and it will return a string describing that error code.

# Contribution Checklist

- [x] first commit title starts with 'ISSUE-N:'
- [x] code approved
- [x] tests approved
- [x] documentation approved
- [x] fixes #20

### Link to gist documentation
https://gist.github.com/desttinghim/01244dd228b48c3c1c028c1da30bbec8

### CODE REVIEW

**Comment by godsflaw** [comment]
I think you need to rebase this off dev and push again.

**Comment by desttinghim** [comment]
Sorry it took so long! I've gotten everything tidied up.

**Comment by godsflaw** [comment]
@desttinghim tests failed in CI
```
make: Entering directory '/node-opendkim/build'                                                                                                        
  CXX(target) Release/obj.target/opendkim/src/opendkim.o                                                                                               
../src/opendkim.cc: In static member function 'static Nan::NAN_METHOD_RETURN_TYPE opendkim::OpenDKIM::SigGetErrorStr(Nan::NAN_METHOD_ARGS_TYPE)':      
../src/opendkim.cc:423:3: error: 'returnstr' was not declared in this scope                                                                            
   returnstr = Nan::New<v8::String>(errorstr).ToLocalChecked();                                                                                        
   ^~~~~~~~~                                                                                                                                           
../src/opendkim.cc:426:16: error: invalid conversion from 'const char**' to 'char**' [-fpermissive]                                                    
     _safe_free(&errorstr);                                                                                                                            
                ^~~~~~~~~                                                                                                                              
In file included from ../src/opendkim.cc:1:0:                                                                                                          
../src/opendkim.h:117:24: note:   initializing argument 1 of 'static void opendkim::OpenDKIM::_safe_free(char**)'                                      
     static inline void _safe_free(char **ptr) {                                                                                                       
                        ^~~~~~~~~~                                                                                                                     
make: *** [opendkim.target.mk:95: Release/obj.target/opendkim/src/opendkim.o] Error 1                                                                  
```

**Comment by godsflaw** [comment]
In the documentation, you also have to fix the example to actually use `sig_geterrorstr()`

**Comment by desttinghim** [comment]
Sorry! Example is fixed and the compile error is resolved.

**Comment by godsflaw** [comment]
@desttinghim the new test file hangs.  I'll look into this tomorrow unless you have an idea.

File: src/opendkim.cc [hunk 1]
```diff
@@ -407,6 +408,24 @@ NAN_METHOD(OpenDKIM::SigGetError) {
...
+  char* errorstr = NULL;
+
+  if (info.Length() != 1) {
+    Nan::ThrowTypeError("sig_geterrorstr(): Wrong number of arguments");
+    goto finish_sig_geterrorstr;
+  }
+
+  errorcode = info[0]->NumberValue(); 
+
+  strcmp(errorstr, dkim_sig_geterrorstr(errorcode));
```
Comment (godsflaw) [inline src/opendkim.cc]: Yeah, so, you never assign `errorstr` here.

Comment (godsflaw) [reply src/opendkim.cc]: the fallout from this is likely why we get a hang.

**Comment by desttinghim** [comment]
You are right! Sorry, my inexperience is probably showing through a little here. I'm confident that it will work this time though!

**Comment by godsflaw** [comment]
@desttinghim there is a problem running the tests that I'm having, not just with this PR, but with another one too.  I'm going to look into that and then push a fix, then you can rebase against it and push one last time.  Sorry for the trouble here, not sure what happened, my guess is some dependency upgraded and broke things.

File: src/opendkim.cc [hunk 2]
```diff
@@ -450,7 +464,8 @@ NAN_METHOD(OpenDKIM::Chunk) {
 
   finish_chunk:
     _safe_free(&message);
-    info.GetReturnValue().Set(info.This());
+
+  info.GetReturnValue().Set(info.This());
```
Comment (godsflaw) [inline src/opendkim.cc:468]: this is cleaner, thanks!

File: src/opendkim.cc [hunk 3]
```diff
@@ -407,6 +408,19 @@ NAN_METHOD(OpenDKIM::SigGetError) {
...
+  if (info.Length() != 1) {                                                     
+    Nan::ThrowTypeError("sig_geterrorstr(): Wrong number of arguments");        
+    return;                                                                     
+  }                                                                             
+                                                                                
+  errorstr = dkim_sig_geterrorstr(info[0]->NumberValue());                                               
+                                                                                
+  info.GetReturnValue().Set(Nan::New<v8::String>(errorstr).ToLocalChecked());   
+}
+
```
Comment (godsflaw) [inline src/opendkim.cc:423]: @desttinghim I made this a little cleaner.  less lines of code, and I got rid of the memory allocation.  If you dig down into `dkim_sig_geterrorstr()` you can see the returned value is just a `const char *` from a mapping table.  No need to get the length, allocate more memory, copy it, and then free it, we can just use that string to pass to our new `v8::String`.  Let me know if you have questions.

File: test/04-verifying/30-sig-geterrorstr.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      message: messages.bad_signature_version,
+      length: messages.bad_signature_version.length
+    });
+    opendkim.chunk_end();
+    t.fail();
+  } catch (err) {
+    var error = opendkim.sig_geterrorstr(opendkim.sig_geterror());
+    t.is(error, 'unsupported signature version');
+  }
+});
```
Comment (godsflaw) [inline test/04-verifying/30-sig-geterrorstr.js:65]: all the tests look good.

File: readme.md [hunk 5]
```diff
@@ -94,9 +94,9 @@ try {
 * [opendkim.get_signature()](https://github.com/godsflaw/node-opendkim/wiki/opendkim.get_signature()): sets the signature info handle.
 * [opendkim.sig_getidentity()](https://github.com/godsflaw/node-opendkim/wiki/opendkim.sig_getidentity()): get the identity from the signature handle.
 * [opendkim.sig_getdomain()](https://github.com/godsflaw/node-opendkim/wiki/opendkim.sig_getdomain()): get the domain from the signature handle.
+* [opendkim.sig_geterror()](https://github.com/godsflaw/node-opendkim/wiki/opendkim.sig_geterror()): Retrieve the error code associated with a rejected/disqualified signature.
```
Comment (godsflaw) [inline readme.md:97]: This text is just lifted from the opendkim docs at:
http://www.opendkim.org/libopendkim/

File: readme.md [hunk 6]
```diff
@@ -94,9 +94,9 @@ try {
...
+* [opendkim.sig_geterrorstr()](https://github.com/godsflaw/node-opendkim/wiki/opendkim.sig_geterrorstr()): get the error string specified by the error code.
```
Comment (godsflaw) [inline readme.md:98]: I adjusted this to be in alphabetical order

**Comment by godsflaw** [comment]
@desttinghim just let me know if these changes look good to you, and I will merge.

**Review by godsflaw — APPROVED** [review]
Looks good, if you think my changes look good.

**Comment by desttinghim** [comment]
The changes look fine to me! Go ahead and merge it.

**Comment by godsflaw** [comment]
@desttinghim I changed the documentation example a little too https://github.com/godsflaw/node-opendkim/wiki/opendkim.sig_geterrorstr()

**Comment by godsflaw** [comment]
@desttinghim feel free to submit your invoice https://ethlance.com/#/job/224

**Pull request merged**